File: scripts/xhs/creator_manage.py
"""创作后台笔记状态和浏览器本地草稿；不删除、不保存、不发布。"""

from __future__ import annotations

import json
import re
import time
from pathlib import Path
from typing import Any

from .errors import XHSError

_SCRIPT = Path(__file__).with_suffix(".js").read_text(encoding="utf-8")
MANAGER_URL = "https://creator.xiaohongshu.com/new/note-manager"
DRAFT_URL = "https://creator.xiaohongshu.com/publish/publish"
KINDS = ("image", "video", "long", "audio")
STATUSES = ("all", "published", "reviewing", "rejected")


class CreatorManageError(XHSError):
    """无法安全读取或定位创作后台内容。"""


def _run(page: Any, action: str, **params: Any) -> dict:
    result = page.evaluate(f"({_SCRIPT})({json.dumps({'action': action, **params})})")
    if not isinstance(result, dict):
        raise CreatorManageError("创作后台没有返回可识别的数据")
    if result.get("error") or result.get("__xhs_error"):
        raise CreatorManageError(result.get("error") or result["__xhs_error"])
    if params.get("account_id") and result.get("ready") is not False:
        _account(result, params["account_id"])
    return result


def _account(result: dict, expected: str = "") -> str:
    account_id = str(result.get("account_id", ""))
    if not re.fullmatch(r"[0-9a-fA-F]{24}", account_id):
        raise CreatorManageError("无法确认当前创作账号身份，未继续读取或打开草稿")
    if expected and account_id != expected:
        raise CreatorManageError("创作账号发生变化，已停止，避免混合账号数据")
    return account_id


def _wait(page: Any, action: str, **params: Any) -> dict:
    deadline = time.monotonic() + 15.0
    while True:
        result = _run(page, action, **params)
        if result.get("ready"):
            return result
        if time.monotonic() >= deadline:
            raise CreatorManageError("页面仍未就绪或结构不受支持；未将加载失败当成空列表")
        time.sleep(0.3)


def _prepare(page: Any, url: str, *, reuse_publish: bool = False) -> None:
    guard = _run(page, "guard")
    if not guard.get("safe_to_navigate"):
        raise CreatorManageError("当前编辑器已有内容，请先保存或处理，未导航或覆盖")
    if not (reuse_publish and guard.get("on_publish")):
        page.navigate(url)
        page.wait_for_load()


def _bounds(limit: int, max_pages: int) -> None:
    if not 1 <= limit <= 500 or not 1 <= max_pages <= 20:
        raise CreatorManageError("limit 必须为 1–500，max-pages 必须为 1–20")


def _collect(page: Any, prefix: str, id_key: str, max_pages: int, **params: Any) -> dict:
    selected = _wait(page, f"{prefix}-select", **params)
    account_id = _account(selected, params.get("account_id", ""))
    params = {**params, "account_id": account_id}
    state = _wait(page, f"{prefix}-read", **params)
    items: dict[str, dict] = {}
    pages = 1
    stop = "end"
    complete = False
    while True:
        _account(state, account_id)
        for item in state["items"]:
            items[item[id_key]] = item
        total = state.get("total")
        complete = bool(state.get("complete")) or (isinstance(total, int) and len(items) >= total)
        if complete:
            break
        if pages >= max_pages:
            stop = "max_pages"
            break
        before = set(items)
        _run(page, f"{prefix}-more", **params)
        deadline = time.monotonic() + 5.0
        while True:
            state = _wait(page, f"{prefix}-read", **params)
            _account(state, account_id)
            if state.get("complete") or any(i[id_key] not in before for i in state["items"]):
                pages += 1
                break
            if time.monotonic() >= deadline:
                stop = "no_progress"
                break
            time.sleep(0.3)
        if stop == "no_progress":
            break
    return {"account_id": account_id, "account_name": state.get("account_name", ""),
            "items": list(items.values()), "complete": complete,
            "total": state.get("total"), "pages": pages, "stop_reason": stop}


def list_managed_notes(
    page: Any, *, status: str = "all", keyword: str = "", limit: int = 100, max_pages: int = 3,
) -> dict:
    _bounds(limit, max_pages)
    if status not in STATUSES:
        raise CreatorManageError("不支持的笔记状态")
    _prepare(page, MANAGER_URL)
    # 读取全部页再筛选已确认的状态，保留未识别状态，避免猜测后台数字枚举。
    result = _collect(page, "notes", "note_id", max_pages, status="all")
    unknown = sum(i["status"] == "unknown" for i in result["items"])
    selected = [i for i in result["items"] if (status == "all" or i["status"] == status)
                and (not keyword or keyword.casefold() in i["title"].casefold())]
    return {**result, "success": True, "scope": "creator_note_manager", "status": status,
            "items": selected[:limit], "count": min(len(selected), limit),
            "total_loaded": len(result["items"]), "unclassified_count": unknown,
            "complete": result["complete"] and len(selected) <= limit
            and (status == "all" or unknown == 0),
            "metrics_scope": "note_manager_snapshot"}


def get_note_status(page: Any, note_id: str, *, max_pages: int = 3) -> dict:
    if not re.fullmatch(r"[0-9a-fA-F]{24}", note_id):
        raise CreatorManageError("笔记 ID 必须为 24 位十六进制 ID")
    result = list_managed_notes(page, limit=500, max_pages=max_pages)
    note = next((i for i in result["items"] if i["note_id"] == note_id), None)
    return {"account_id": result["account_id"], "account_name": result.get("account_name", ""),
            "success": note is not None, "found": note is not None,
            "note": note, "search_complete": result["complete"],
            "scope": "creator_note_manager",
            "message": "已找到笔记" if note else "已读范围内未找到；不据此判断笔记已删除"}


def list_drafts(
    page: Any, *, kind: str = "all", keyword: str = "", limit: int = 100, max_pages: int = 3,
) -> dict:
    _bounds(limit, max_pages)
    if kind not in (*KINDS, "all"):
        raise CreatorManageError("不支持的草稿类型")
    _prepare(page, DRAFT_URL, reuse_publish=True)
    drawer = _wait(page, "draft-drawer")
    account_id = _account(drawer)
    results = [{**_collect(page, "draft", "draft_id", max_pages, kind=k, account_id=account_id),
                "kind": k}
               for k in (KINDS if kind == "all" else (kind,))]
    items = [i for result in results for i in result["items"]]
    selected = [i for i in items if not keyword or keyword.casefold() in i["title"].casefold()]
    return {"account_id": account_id, "account_name": drawer.get("account_name", ""),
            "success": True, "scope": "current_browser_local_drafts", "kind": kind,
            "items": selected[:limit], "count": min(len(selected), limit),
            "total_loaded": len(items),
            "complete": all(r["complete"] for r in results) and len(selected) <= limit,
            "groups": [{k: v for k, v in r.items() if k != "items"} for r in results]}


def open_draft(
    page: Any, draft_id: str, expected_title: str, kind: str, *, open_editor: bool = False,
    max_pages: int = 3,
) -> dict:
    if not re.fullmatch(r"[A-Za-z0-9_-]{1,100}", draft_id) or not expected_title.strip():
        raise CreatorManageError("必须提供稳定的草稿 ID 和完整显示标题")
    if kind not in KINDS:
        raise CreatorManageError("打开草稿时必须指定具体类型")
    result = list_drafts(page, kind=kind, limit=500, max_pages=max_pages)
    matches = [i for i in result["items"] if i["draft_id"] == draft_id]
    if len(matches) != 1 or matches[0]["title"] != expected_title:
        raise CreatorManageError("已读取的草稿中 ID 和标题未同时匹配，未打开")
    account_id = _account(result)
    base = {"account_id": account_id, "account_name": result.get("account_name", ""),
            "draft": matches[0], "saved": False, "published": False}
    if not open_editor:
        return {**base, "success": True, "status": "preview", "opened": False}
    # ID、标题与已有编辑内容检查在同一次页面操作中完成。
    try:
        clicked = _run(page, "draft-open", kind=kind, draft_id=draft_id,
                       expected_title=expected_title, account_id=account_id)
        if not clicked.get("clicked"):
            raise CreatorManageError("未确认草稿编辑按钮已点击")
        state = _wait(page, "editor", account_id=account_id)
        _account(state, account_id)
        wanted = expected_title if matches[0].get("has_title") else ""
        if wanted not in state.get("titles", []):
            raise CreatorManageError("打开后的编辑器标题与草稿不一致")
        return {**base, "success": True, "status": "editor_open", "opened": True}
    except Exception:
        return {**base, "success": False, "status": "unknown", "opened": None,
                "error": "未能确认编辑器已打开目标草稿，请检查当前页面；未保存或发布"}

File: scripts/xhs/creator_manage.js
(params) => {
    const visible = el => !!el && el.getBoundingClientRect().width > 0 &&
        el.getBoundingClientRect().height > 0 && getComputedStyle(el).visibility !== 'hidden' &&
        Number(getComputedStyle(el).opacity || 1) > 0.1;
    const all = (selector, root = document) => [...root.querySelectorAll(selector)].filter(visible);
    const text = el => el?.textContent.trim() || '';
    const plain = v => JSON.parse(JSON.stringify(v));
    const creator = location.origin === 'https://creator.xiaohongshu.com';
    const onPublish = creator && /^\/publish\//.test(location.pathname);
    const drawer = all('.d-drawer').find(el => el.querySelector('.draft-list')) ||
        all('.draft-list')[0]?.closest('.d-drawer-container') || all('.draft-list')[0]?.parentElement;
    const outsideDraftList = el => !drawer?.contains(el) &&
        !el.closest('.draft-drawer, .draft-list, .draft-item');
    const controls = all('input:not([type="file"]), textarea, [contenteditable="true"]')
        .filter(outsideDraftList);
    const titles = controls.filter(el => el.tagName === 'INPUT' || el.tagName === 'TEXTAREA')
        .map(el => String(el.value || '').trim());
    const editors = controls.filter(el => el.getAttribute('contenteditable') === 'true');
    const dirty = onPublish && (controls.some(el =>
        String(el.value || el.textContent || '').trim() || el.querySelector('img,video,audio')) ||
        all('.img-preview-area .pr, .video-preview, video[src], audio[src]').some(outsideDraftList) ||
        [...document.querySelectorAll('input[type="file"]')]
            .some(el => outsideDraftList(el) && el.files?.length));
    if (params.action === 'guard') return {safe_to_navigate: !dirty, on_publish: onPublish};
    if (!creator) return {error: '当前不是小红书创作服务平台'};
    if (all('.login-container, .login-page').length) return {error: '创作服务平台尚未登录'};
    const app = document.querySelector('#app')?.__vue_app__;
    const unwrap = value => value && typeof value === 'object' && 'value' in value ? value.value :
        value && typeof value === 'object' && '_value' in value ? value._value : value;
    const user = unwrap(app?.config?.globalProperties?.$store?.state?.Auth?.userInfo);
    const account = {account_id: String(user?.userId || user?.user_id || ''),
        account_name: String(user?.userName || user?.nickname || '')};
    if (!/^[0-9a-fA-F]{24}$/.test(account.account_id)) return {ready: false};
    if (params.account_id && params.account_id !== account.account_id) {
        return {...account, error: '创作账号发生变化，已停止，避免混合账号数据或打开错误草稿'};
    }
    const result = (() => {
    if (params.action === 'editor') return {ready: onPublish && !drawer &&
        (editors.length > 0 || all('.img-preview-area .pr, .video-preview').length > 0), titles};
    if (params.action.startsWith('draft')) {
        if (!onPublish) return {error: '当前不是发布页面'};
        if (params.action === 'draft-drawer') {
            if (drawer) return {ready: true};
            if (dirty) return {error: '编辑器已有内容，请先保存或处理，避免覆盖'};
            const entries = all('.draft-title').filter(el => /^草稿箱/.test(text(el)));
            if (entries.length !== 1) return {ready: false};
            entries[0].click();
            return {ready: false};
        }
        if (!drawer) return {ready: false};
        const labels = {image: '图文笔记', video: '视频笔记', long: '长文笔记', audio: '播客笔记'};
        const tab = all('.tab-item', drawer).find(el => text(el).startsWith(labels[params.kind]));
        if (!tab) return {error: '草稿箱没有请求的类型标签'};
        if (params.action === 'draft-select') {
            if (!tab.classList.contains('active')) tab.click();
            return {ready: true};
        }
        if (!tab.classList.contains('active')) return {ready: false};
        const count = Number(text(tab).match(/[（(](\d+)[）)]/)?.[1] ?? NaN);
        const cards = all('.draft-item[data-draft-id]', drawer);
        const items = cards.map(el => ({draft_id: el.getAttribute('data-draft-id'),
            kind: params.kind, raw_type: el.getAttribute('data-draft-type'),
            title: text(el.querySelector('.draft-title-text')),
            saved_at: text(el.querySelector('.draft-time')),
            has_title: el.querySelector('.draft-title-text')?.classList.contains('has-title') || false}));
        if (params.action === 'draft-open') {
            if (dirty) return {error: '编辑器已有内容，未覆盖'};
            const matches = cards.filter(el => el.getAttribute('data-draft-id') === params.draft_id);
            if (matches.length !== 1 || text(matches[0].querySelector('.draft-title-text')) !== params.expected_title) {
                return {error: '草稿 ID 和标题未同时匹配，未打开'};
            }
            const buttons = all('.draft-actions .btn', matches[0]).filter(el => text(el) === '编辑');
            if (buttons.length !== 1) return {error: '未找到唯一的草稿编辑按钮'};
            buttons[0].click();
            return {clicked: true};
        }
        if (params.action === 'draft-more') {
            const list = drawer.querySelector('.draft-list');
            list.scrollTo({top: list.scrollHeight, behavior: 'instant'});
            list.dispatchEvent(new Event('scroll', {bubbles: true}));
            return {scrolled: true};
        }
        const empty = all('.empty-state', drawer).some(el => /暂无草稿/.test(text(el)));
        if (!items.length && !(count === 0 && empty)) return {ready: false};
        return plain({ready: true, items, total: Number.isFinite(count) ? count : null,
            complete: Number.isFinite(count) && items.length >= count});
    }
    if (location.pathname !== '/new/note-manager') return {error: '当前不是笔记管理页面'};
    const container = document.querySelector('.notes-container');
    if (!container) return {ready: false};
    const tabs = all('.tab-list .tab-item', container);
    const labels = {all: '全部', published: '已发布', reviewing: '审核中', rejected: '未通过'};
    const tab = tabs.find(el => text(el).startsWith(labels[params.status]));
    if (!tab) return {ready: false};
    if (params.action === 'notes-select') {
        if (!tab.classList.contains('tab-item--active')) tab.click();
        return {ready: true};
    }
    if (!tab.classList.contains('tab-item--active')) return {ready: false};
    if (all('.d-loading, .loading', container).length) return {ready: false};
    const errors = all('.error-message, .load-error', container).map(text).filter(Boolean);
    if (errors.length) return {error: errors.join('；')};
    if (params.action === 'notes-more') {
        const next = all('button, [role="button"]', container).find(el =>
            el.getAttribute('aria-label') === '下一页' || text(el) === '下一页');
        if (next && !next.disabled && next.getAttribute('aria-disabled') !== 'true') next.click();
        else {
            const scrollables = [container, ...container.querySelectorAll('*')].filter(el =>
                visible(el) && el.scrollHeight > el.clientHeight + 8 &&
                /auto|scroll/.test(getComputedStyle(el).overflowY));
            const scroller = scrollables.at(-1) || document.scrollingElement;
            scroller.scrollTo({top: scroller.scrollHeight, behavior: 'instant'});
            scroller.dispatchEvent(new Event('scroll', {bubbles: true}));
        }
        return {scrolled: true};
    }
    const props = new Map();
    const seen = new Set();
    function walk(v, depth = 0) {
        if (!v || typeof v !== 'object' || seen.has(v) || depth > 70) return;
        seen.add(v);
        if (v.component) {
            const c = v.component;
            const note = c.props?.note;
            if (note?.id) props.set(String(note.id), note);
            walk(c.subTree, depth + 1);
        }
        if (Array.isArray(v.children)) v.children.forEach(x => walk(x, depth + 1));
    }
    walk(document.querySelector('#app')?.__vue_app__?._container?._vnode);
    const cards = all('.note-card', container);
    const items = [];
    for (const card of cards) {
        let id;
        try {id = JSON.parse(card.dataset.impression || '{}').noteTarget?.value?.noteId;} catch (_) {}
        if (!id) return {error: '笔记卡片缺少稳定 ID，未输出可能错配的记录'};
        const note = props.get(id);
        const title = text(card.querySelector('.note-card__title'));
        if (note && String(note.display_title) !== title) return {ready: false};
        items.push({note_id: id, title, time: note?.time || text(card.querySelector('.note-card__time')),
            status: note?.tab_status === 1 ? 'published' :
                all('.note-card__status, .note-card__tag', card).some(el => text(el) === '审核中') ? 'reviewing' :
                all('.note-card__status, .note-card__tag', card).some(el => text(el) === '未通过') ? 'rejected' : 'unknown',
            raw_status: note?.tab_status ?? null, type: note?.type || 'unknown',
            xsec_token: note?.xsec_token || '', stats: note ? {views: note.view_count ?? null,
                likes: note.likes ?? null, comments: note.comments_count ?? null,
                collections: note.collected_count ?? null, shares: note.shared_count ?? null} : null});
    }
    const totalText = text(tabs.find(el => text(el).startsWith('全部')));
    const allTotal = Number(totalText.match(/全部\s*[（(]?\s*(\d+)/)?.[1] ?? NaN);
    const empty = all('.empty-state, .d-result, .empty', container).some(el => /暂无|没有.*笔记/.test(text(el)));
    if (!items.length && !empty && allTotal !== 0) return {ready: false};
    const complete = (Number.isFinite(allTotal) && items.length >= allTotal) || empty;
    return plain({ready: true, items, total: Number.isFinite(allTotal) ? allTotal : null,
        complete, status: params.status});
    })();
    return plain({...result, ...account});
}
